Sequence Updates

lhirata opened the pull request:
Closes #2718, #2510, #2516, #2522.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/2779)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f4478e873c6b9a41f859fb4ba1f4eeaade8e5833/

vilterp commented on v2.0/rename-sequence.md:
```diff
@@ -4,9 +4,9 @@ summary: The RENAME SEQUENCE statement changes the name of a sequence.
+<span class="version-tag">New in v2.0:</span> The `RENAME TO` [statement](sql-statements.html) is part of [`ALTER SEQUENCE`](alter-sequence.html), and changes the name of a sequence.
 
-{{site.data.alerts.callout_danger}}If you rename a sequence that's being used in a table, the change will not propagate to the <code>DEFAULT</code> expressions that reference the sequence. {{site.data.alerts.end}}
+{{site.data.alerts.callout_danger}}You cannot rename a sequence that's being used in a table. To rename the sequence, drop the <code>DEFAULT</code> expressions that reference the sequence, rename the sequence, and add the <code>DEFAULT</code> expressions back.{{site.data.alerts.end}}
```
👍

vilterp replied:
May want to mention or link to the syntax for doing this: `ALTER TABLE _ ALTER COLUMN _ DROP DEFAULT`

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/7b9ebd7ed0812c104759848d816d2897c727d96a/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/a5908f05c454295e9aac2733c44e7d2ed26c883f/

vilterp commented on v2.0/sql-dump.md:
```diff
@@ -14,8 +14,8 @@ The `cockroach dump` [command](cockroach-commands.html) outputs the SQL statemen
-- Table and view schemas and table data are dumped as they appeared at the time that the command is started. Any changes after the command starts will not be included in the dump.
-- **New in v1.1:** Table and view schemas are dumped in the order in which they can successfully be recreated.
+- Table, sequence, and view schemas and table data are dumped as they appeared at the time that the command is started. Any changes after the command starts will not be included in the dump.
+- **New in v1.1:** Table, sequence, and view schemas are dumped in the order in which they can successfully be recreated.
```
Sadly not true if you have cyclical foreign keys: https://github.com/cockroachdb/cockroach/issues/20254

Worth noting as a known limitation if it's not already.

vilterp commented on v2.0/create-sequence.md:
```diff
@@ -43,6 +43,7 @@ table td:first-child {
 We support the following [SQL sequence functions](functions-and-operators.html):
 
 - `nextval('seq_name')`
+    {{site.data.alerts.callout_info}}If <code>nextval()</code> is used in conjunction with <code>RETURNING NOTHING</code> statements, the sequence increments can be reordered. For more information, see <a href="parallel-statement-execution.html">Parallel Statement Execution</a>.{{site.data.alerts.end}}
```
👍

vilterp commented on v2.0/backup.md:
```diff
@@ -26,9 +26,10 @@ Dependent objects should be backed up at the same time as the objects they depen
-[Views](views.html) | The tables used in the view's `SELECT` statement
-[Interleaved tables](interleave-in-parent.html) | The parent table in the [interleaved hierarchy](interleave-in-parent.html#interleaved-hierarchy)
+Table with [foreign key](foreign-key.html) constraints | The table it `REFERENCES` (however, this dependency can be [removed during the restore](restore.html#skip_missing_foreign_keys)).
+Table with a [sequence](create-sequence.html) | The sequence.
```
There is actually an option `skip_missing_sequences` which works the same way as `skip_missing_foreign_keys`. Sorry for not calling it out better in release notes or otherwise! I think we need to add something to `restore.md` and link to it here, same as for `skip_missing_foreign_keys`.

Mimicking the section for foreign keys, the new section in `restore.md` would be something like:

> #### `skip_missing_sequences`
>
> - **Description**: If you want to restore a table that depends on a sequence but don't want to restore the sequence it references, you can drop the sequence dependency from a table (its sequence-using `DEFAULT` expression) and then have it restored.
> - **Key**: `skip_missing_sequences`
> - **Value**: *No value*
> - **Example**: `WITH skip_missing_sequences`

(Note: it's a bit odd that the `skip_missing_foreign_keys` section of that doc seems to link to its own heading; removed that)

lhirata commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[v2.0/backup.md, line 30 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2779#-L8XhmSBX4VCHFRKigr9-r3-30:-L8baJ2Kfm4GljlLY3de:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/a5908f05c454295e9aac2733c44e7d2ed26c883f/v2.0/backup.md#L30)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

There is actually an option `skip_missing_sequences` which works the same way as `skip_missing_foreign_keys`. Sorry for not calling it out better in release notes or otherwise! I think we need to add something to `restore.md` and link to it here, same as for `skip_missing_foreign_keys`.

Mimicking the section for foreign keys, the new section in `restore.md` would be something like:

> #### `skip_missing_sequences`
>
> - **Description**: If you want to restore a table that depends on a sequence but don't want to restore the sequence it references, you can drop the sequence dependency from a table (its sequence-using `DEFAULT` expression) and then have it restored.
> - **Key**: `skip_missing_sequences`
> - **Value**: *No value*
> - **Example**: `WITH skip_missing_sequences`

(Note: it's a bit odd that the `skip_missing_foreign_keys` section of that doc seems to link to its own heading; removed that)
</blockquote></details>

Done.

---

*[v2.0/rename-sequence.md, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2779#-L8XhmSBX4VCHFRKigrB-r2-9:-L8bb2BfnAns3Rerk5WX:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/7b9ebd7ed0812c104759848d816d2897c727d96a/v2.0/rename-sequence.md#L9)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

May want to mention or link to the syntax for doing this: `ALTER TABLE _ ALTER COLUMN _ DROP DEFAULT`
</blockquote></details>

Done.

---

*[v2.0/sql-dump.md, line 18 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2779#-L8XhmSBX4VCHFRKigrD-r3-18:-L8bcSnW2zYQ7AH9gUGd:b8yxl47) ([raw file](https://github.com/cockroachdb/docs/blob/a5908f05c454295e9aac2733c44e7d2ed26c883f/v2.0/sql-dump.md#L18)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Sadly not true if you have cyclical foreign keys: https://github.com/cockroachdb/cockroach/issues/20254

Worth noting as a known limitation if it's not already.
</blockquote></details>

Added clarification. Also tagged the bug ticket with "known limitation" so we can document it in the aggregated list.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2779)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e7c41ad9305e8631444f84e69c629108169eb59b/

jseldess commented on v2.0/restore.md:
```diff
@@ -130,7 +131,14 @@ You can include the following options as key-value pairs in the `kv_option_list`
+#### `skip_missing_sequences` <span class="version-tag">New in v2.0</span>
+
+- **Description**: If you want to restore a table that depends on a sequence but don't want to restore the sequence it references, you can drop the sequence dependency from a table (i.e., the `DEFAULT` expression that uses the sequence) and then have it restored.
 - **Key**: `skip_missing_foreign_keys`
```
Should the key/value/example be different for sequences?

jseldess commented on v2.0/backup.md:
```diff
@@ -26,9 +26,10 @@ Dependent objects should be backed up at the same time as the objects they depen
-[Views](views.html) | The tables used in the view's `SELECT` statement
-[Interleaved tables](interleave-in-parent.html) | The parent table in the [interleaved hierarchy](interleave-in-parent.html#interleaved-hierarchy)
+Table with [foreign key](foreign-key.html) constraints | The table it `REFERENCES` (however, this dependency can be [removed during the restore](restore.html#skip_missing_foreign_keys)).
+Table with a [sequence](create-sequence.html) | <span class="version-tag">New in v2.0:</span> The sequence it uses (however, this dependency can be [removed during the restore](restore.html#skip_missing_sequences-new-in-v2-0).
```
nit: In this and the preceding row, could you change the parenthetical to a stand-alone sentence (or separate with a semi-colon)?

jseldess commented on v2.0/restore.md:
```diff
@@ -130,7 +131,14 @@ You can include the following options as key-value pairs in the `kv_option_list`
+- **Value**: *No value*
+- **Example**: `WITH skip_missing_foreign_keys`
+
+#### `skip_missing_sequences` <span class="version-tag">New in v2.0</span>
```
Unfortunately, the callout doesn't render correctly for h4 and up. Would you mind filing this bug with @j-low and, for the time being, just moving this to the line below the heading?

lhirata commented:
Review status: 0 of 6 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[v2.0/backup.md, line 30 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2779#-L8XhmSBX4VCHFRKigr9-r4-30:-L8cKrDIs9Ex4aSbrOHo:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e7c41ad9305e8631444f84e69c629108169eb59b/v2.0/backup.md#L30)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

nit: In this and the preceding row, could you change the parenthetical to a stand-alone sentence (or separate with a semi-colon)?
</blockquote></details>

Done.

---

*[v2.0/restore.md, line 139 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2779#-L8XhmSBX4VCHFRKigrC-r4-139:-L8cLNjdA-tTaThqrZZ3:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e7c41ad9305e8631444f84e69c629108169eb59b/v2.0/restore.md#L139)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Unfortunately, the callout doesn't render correctly for h4 and up. Would you mind filing this bug with @j-low and, for the time being, just moving this to the line below the heading? 
</blockquote></details>

Done.

---

*[v2.0/restore.md, line 142 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2779#-L8XhmSBX4VCHFRKigrC-r4-142:-L8cLRfaDf62jtPhe52m:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e7c41ad9305e8631444f84e69c629108169eb59b/v2.0/restore.md#L142)):*
<details><summary><i>Previously, jseldess (Jesse Seldess) wrote…</i></summary><blockquote>

Should the key/value/example be different for sequences?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2779)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8616f7362f0db17f629ee2a350621e4647b93ef3/

lhirata commented:
TFTRs! 🎉
